src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,122 @@ impl<S: StateStore> StateTable<S> {
+        StateTableRowIter::new(
+            &self.keyspace,
+            &self.column_descs,
+            mem_table_iter,
+            &self.order_types,
+            epoch,
+        )
+        .await
     }
 }
 
-pub struct StateTableRowIter<S: StateStore> {
-    cell_based_iter: CellBasedTableRowIter<S>,
-    mem_table: MemTable,
+/// `StateTableRowIter` is able to read the just written data (uncommited data).
+/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`
+pub struct StateTableRowIter<'a, S: StateStore> {
+    mem_table_iter: Peekable<MemTableIter<'a>>,
+    cell_based_streaming_iter: CellBasedTableStreamingIter<S>,
+
+    /// The result of the last cell_based_streaming_iter next is saved, which can avoid being
+    /// discarded.
+    cell_based_item: Option<(Vec<u8>, Row)>,
+    pk_serializer: OrderedRowSerializer,
 }
+type MemTableIter<'a> = btree_map::Iter<'a, Row, RowOp>;
 
-impl<S: StateStore> StateTableRowIter<S> {
+impl<'a, S: StateStore> StateTableRowIter<'a, S> {
     async fn new(
-        _keyspace: Keyspace<S>,
-        _table_descs: Vec<ColumnDesc>,
-        _epoch: u64,
-        _stats: Arc<StateStoreMetrics>,
-    ) -> StorageResult<Self> {
-        todo!()
+        keyspace: &Keyspace<S>,
+        table_descs: &[ColumnDesc],
+        mem_table_iter: Peekable<MemTableIter<'a>>,
+        order_types_vec: &[OrderType],
+        epoch: u64,
+    ) -> StorageResult<StateTableRowIter<'a, S>> {
+        let mut cell_based_streaming_iter =
+            CellBasedTableStreamingIter::new(keyspace, table_descs, epoch).await?;
+        let pk_serializer = OrderedRowSerializer::new(order_types_vec.to_vec());
+        let cell_based_item = cell_based_streaming_iter.next().await.map_err(err)?;
+        let state_table_iter = Self {
+            mem_table_iter,
+            cell_based_streaming_iter,
+            cell_based_item,
+            pk_serializer,
+        };
+
+        Ok(state_table_iter)
     }
 
-    async fn next(&mut self) -> StorageResult<Option<Row>> {
-        todo!()
+    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        let mut mem_table_iter_next_flag = false;
+        let mut cell_based_iter_next_flag = false;
+        let mut res = None;
+        let cell_based_item = std::mem::take(&mut self.cell_based_item);
+        match (cell_based_item, self.mem_table_iter.peek()) {
+            (None, None) => {
+                res = None;
+            }
+            (Some((_, row)), None) => {
+                res = Some(row);
+                cell_based_iter_next_flag = true;
+            }
+            (None, Some((_, row_op))) => {
+                mem_table_iter_next_flag = true;
+                match row_op {
+                    RowOp::Insert(row) => res = Some(row.clone()),
+                    RowOp::Delete(_) => {}
+                    RowOp::Update((_, new_row)) => res = Some(new_row.clone()),
```

Comment:
We should avoid these `clone`s by calling `self.mem_table_iter.next()` here.

Reply:
My fault. The `mem_table_iter` is actually a borrowed iterator, so there's no way to avoid this `clone` if we want an owned `Row` here.

However, from another point of view, do the callers always want the owned row from the iterator? Maybe returning a `Cow<Row>` will be better?